File: examples/python/basic_usage.py
"""
UTF-PARC Example (Python)
=========================

Minimal example showing how to use the PARC core library:

- encode_text()   → turn free text into a (c, m, f, k) vector
- update_state()  → apply temporal update steps
- validate_parc() → sanity-check a vector

This script assumes the repo layout:

parc-protocol/
  src/
    python/
      parc.py
  examples/
    python/
      basic_usage.py

You can run it with:

    python examples/python/basic_usage.py
"""

import os
import sys
import json

# ---------------------------------------------------------------------
# Make sure we can import src/python/parc.py without installing a package
# ---------------------------------------------------------------------

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
SRC_PYTHON_DIR = os.path.join(ROOT_DIR, "src", "python")

if SRC_PYTHON_DIR not in sys.path:
    sys.path.insert(0, SRC_PYTHON_DIR)

from parc import encode_text, update_state, validate_parc  # type: ignore


def pretty(obj) -> str:
    """Small helper for nicely formatted JSON output."""
    return json.dumps(obj, indent=2, sort_keys=True)


def main() -> None:
    print("=== UTF-PARC Python Example ===\n")

    # -----------------------------------------------------------------
    # 1. Encode a piece of text into a PARC vector
    # -----------------------------------------------------------------
    text = "Congress makes laws and the President signs them into law."

    print("Input text:")
    print(f"  {text}\n")

    state = encode_text(text)
    print("Initial PARC state (c, m, f, k):")
    print(pretty(state))
    print()

    # -----------------------------------------------------------------
    # 2. Validate the vector
    # -----------------------------------------------------------------
    validation = validate_parc(state)
    print("Validation result:")
    print(pretty(validation))
    print()

    # -----------------------------------------------------------------
    # 3. Apply temporal update steps (simulating learning over turns)
    # -----------------------------------------------------------------
    params = {
        "gamma": 0.3,   # learning rate for correctness
        "delta": 0.1,   # decay rate for misconceptions
        "beta": 0.5,    # interaction: correctness vs misconception
        "rho": 0.5,     # confidence growth rate
    }

    updated = update_state(state, steps=3, params=params)
    print("Updated PARC state after 3 steps:")
    print(pretty(updated))
    print()

    # -----------------------------------------------------------------
    # 4. Show key interpretations
    # -----------------------------------------------------------------
    c = updated["c"]
    m = updated["m"]
    f = updated["f"]
    k = updated["k"]

    print("Interpretation:")
    print(f"  correctness (c):   {c:.3f}")
    print(f"  misconception (m): {m:.3f}")
    print(f"  fog (f):           {f:.3f}")
    print(f"  confidence (k):    {k:.3f}")
    print()

    if m > 0.3 and k > 0.6:
        print("→ High-risk: confidently wrong (high m, high k).")
    elif f > 0.4:
        print("→ High fog: learner is uncertain / incomplete.")
    elif c > 0.7 and m < 0.2:
        print("→ Strong understanding with low misconception.")
    else:
        print("→ Mixed state: partial understanding with some noise.")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: examples/js/basic_usage.mjs
/**
 * UTF-PARC Example (JavaScript / Node.js)
 * =======================================
 *
 * Demonstrates how to use the JS PARC module:
 *
 *   import { encodeText, updateState, validateParc } from "../../src/js/parc.mjs";
 *
 * Run with:
 *
 *     node examples/js/basic_usage.mjs
 */

import { dirname, resolve } from "path";
import { fileURLToPath } from "url";

// ---------------------------------------------------------
// Resolve path to src/js/parc.mjs so we can import locally
// ---------------------------------------------------------
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const PARC_PATH = resolve(__dirname, "../../src/js/parc.mjs");

// Dynamic import (works everywhere)
const { encodeText, updateState, validateParc } = await import(PARC_PATH);

// ---------------------------------------------------------
// Helper for nicer JSON printing
// ---------------------------------------------------------
function pretty(obj) {
  return JSON.stringify(obj, null, 2);
}

// ---------------------------------------------------------
// Main Example
// ---------------------------------------------------------
console.log("=== UTF-PARC JS Example ===\n");

// 1. Encode text
const text = "Photosynthesis converts light energy into chemical energy.";
console.log("Input text:");
console.log(" ", text, "\n");

const state = encodeText(text);
console.log("Initial PARC state (c, m, f, k):");
console.log(pretty(state), "\n");

// 2. Validate the vector
const validation = validateParc(state);
console.log("Validation:");
console.log(pretty(validation), "\n");

// 3. Apply temporal update steps
const params = {
  gamma: 0.3,
  delta: 0.1,
  beta: 0.5,
  rho: 0.5
};

const updated = updateState(state, 3, params);
console.log("Updated PARC state after 3 steps:");
console.log(pretty(updated), "\n");

// 4. Interpretation
const { c, m, f, k } = updated;

console.log("Interpretation:");

if (m > 0.3 && k > 0.6) {
  console.log(" → High-risk: confidently wrong (high m, high k).");
} else if (f > 0.4) {
  console.log(" → High fog: uncertainty or incomplete reasoning.");
} else if (c > 0.7 && m < 0.2) {
  console.log(" → Strong understanding with low misconception.");
} else {
  console.log(" → Mixed state.");
}

console.log("\nDone.");
